lightgbm/src/main/scala/com/microsoft/ml/spark/lightgbm/dataset/DatasetAggregator.scala

```diff
@@ -0,0 +1,563 @@
+          chunkSize)
+    }
+    initScoreChunkedArrayOpt.foreach {
+      initScoreChunkedArray =>
+        ChunkedArrayUtils.copyChunkedArray(initScoreChunkedArray, this.initScoreArrayOpt.get,
+          threadInitScoreStartIndex, chunkSize)
+    }
+    ChunkedArrayUtils.copyChunkedArray(indexesChunkedArray, this.indexesArray, threadIndexesStartIndex, chunkSize)
+    ChunkedArrayUtils.copyChunkedArray(valuesChunkedArray, this.valuesArray, threadIndexesStartIndex, chunkSize)
+    ChunkedArrayUtils.copyChunkedArray(indptrChunkedArray, this.indptrArray, threadIndPtrStartIndex, chunkSize)
+    groupColumnValues.copyToArray(groupColumnValuesArray, threadRowStartIndex.toInt)
+    // rewrite array reference for volatile arrays, see: https://www.javamex.com/tutorials/volatile_arrays.shtml
+    this.synchronized {
+      groupColumnValuesArray = groupColumnValuesArray
+    }
+  }
+
+  /** Adds the rows to the internal data structure.
+    * @param labelsChunkedArray The column of label values.
+    * @param weightChunkedArrayOpt The optional column of weights, if specified.
+    * @param initScoreChunkedArrayOpt The optional column of initial scores, if specified.
+    * @param indexesChunkedArray The feature SparseVector indexes.
+    * @param valuesChunkedArray The feature SparseVector values.
+    * @param indptrChunkedArray The feature SparseVector indptr.
+    * @param groupColumnValues The column of group values, if in ranking scenario.
+    */
+  def addRows(labelsChunkedArray: FloatChunkedArray,
+              weightChunkedArrayOpt: Option[FloatChunkedArray],
+              initScoreChunkedArrayOpt: Option[DoubleChunkedArray],
+              indexesChunkedArray: IntChunkedArray,
+              valuesChunkedArray: DoubleChunkedArray,
+              indptrChunkedArray: IntChunkedArray,
+              groupColumnValues: ListBuffer[Row]): Unit = {
+    parallelInitializeRows(weightChunkedArrayOpt, initScoreChunkedArrayOpt)
+    parallelizedCopy(labelsChunkedArray, weightChunkedArrayOpt, initScoreChunkedArrayOpt, indexesChunkedArray,
+      valuesChunkedArray, indptrChunkedArray, groupColumnValues)
+  }
+}
+
+private[lightgbm] final class DatasetAggregator(columnParams: ColumnParams,
+                                                schema: StructType,
+                                                useSingleDataset: Boolean,
+                                                chunkSize: Int,
+                                                matrixType: String,
+                                                val sharedState: SharedState)
+  extends LocalAggregator[AggregatedColumns] {
+  lazy val denseConsolidatorHolder = SharedSingleton { new DenseSyncAggregatedColumns(chunkSize) }
+
+  lazy val sparseConsolidatorHolder = SharedSingleton { new SparseSyncAggregatedColumns(chunkSize) }
+
+  def prep(iter: Iterator[Row]): AggregatedColumns = {
+    var (concatRowsIter: Iterator[Row], isSparse: Boolean) = getArrayType(iter,
```

Comment:
could be a val

Reply:
the isSparse is reassigned so it has to be a var, do you see a way to rewrite it to make the concatRowsIter a val?  I can't seem to make the tuple contain both a var and val.